synapse/rest/admin/users.py

```diff
@@ -225,6 +225,37 @@ async def on_PUT(
                         user_id, threepid["medium"], threepid["address"], current_time
                     )
 
+            if "external_ids" in body:
+                # check for required parameters for each external_id
+                # and convert into List[Tuple[str, str]]
+                new_external_ids = []
+                for external_id in body["external_ids"]:
+                    assert_params_in_dict(external_id, ["auth_provider", "external_id"])
+                    new_external_ids.append(
+                        (external_id["auth_provider"], external_id["external_id"])
+                    )
+
+                # get changed external_ids (added and removed)
+                cur_external_ids = await self.store.get_external_ids_by_user(user_id)
+                add_external_ids = set(new_external_ids) - set(cur_external_ids)
+                del_external_ids = set(cur_external_ids) - set(new_external_ids)
+
+                # remove old external_ids
+                for auth_provider, external_id in del_external_ids:
+                    await self.store.remove_user_external_id(
+                        auth_provider,
+                        external_id,
+                        user_id,
+                    )
+
+                # add new external_ids
+                for auth_provider, external_id in add_external_ids:
+                    await self.store.record_user_external_id(
+                        auth_provider,
+                        external_id,
+                        user_id,
+                    )
```

Comment:
The question is what is the better way to update the IDs?
- Remove all IDs and then add the given ID? This is the way for `threepids`. When you add a additional ID, all IDs become removed and afterwards added. This can cause an ID to be removed and added even though it is not affected by the change.
- Determine which ID is to be added or removed and write/remove only this to the database. This is the what I do with `external_ids`.

I would like prefer the same kind of implementation for both, `threepid` and `external_ids`. IMO it is good when the code for `threepid` and `external_ids` is very similar to each other.

Reply:
Ah, I see. Yes, I think your approach is better, since it means fewer updates to the database.